command/agent/job_endpoint.go

```diff
@@ -1007,39 +1007,50 @@ func ApiServicesToStructs(in []*api.Service) []*structs.Service {
 			}
 		}
 
-		if s.Connect == nil {
-			continue
+		if s.Connect != nil {
+			out[i].Connect = ApiConsulConnectToStructs(s.Connect)
 		}
 
-		out[i].Connect = &structs.ConsulConnect{
-			Native: s.Connect.Native,
-		}
+	}
 
-		if s.Connect.SidecarService == nil {
-			continue
-		}
+	return out
+}
 
-		out[i].Connect.SidecarService = &structs.ConsulSidecarService{
-			Port: s.Connect.SidecarService.Port,
-		}
+func ApiConsulConnectToStructs(in *api.ConsulConnect) *structs.ConsulConnect {
+	if in == nil {
+		return nil
+	}
 
-		if s.Connect.SidecarService.Proxy == nil {
-			continue
-		}
+	out := &structs.ConsulConnect{
+		Native: in.Native,
+	}
+
+	if in.SidecarService != nil {
 
-		out[i].Connect.SidecarService.Proxy = &structs.ConsulProxy{
-			Config: s.Connect.SidecarService.Proxy.Config,
+		out.SidecarService = &structs.ConsulSidecarService{
+			Port: in.SidecarService.Port,
 		}
 
-		upstreams := make([]*structs.ConsulUpstream, len(s.Connect.SidecarService.Proxy.Upstreams))
-		for i, p := range s.Connect.SidecarService.Proxy.Upstreams {
-			upstreams[i] = &structs.ConsulUpstream{
-				DestinationName: p.DestinationName,
-				LocalBindPort:   p.LocalBindPort,
+		if in.SidecarService.Proxy != nil {
+
+			out.SidecarService.Proxy = &structs.ConsulProxy{
+				Config: in.SidecarService.Proxy.Config,
 			}
+
+			upstreams := make([]*structs.ConsulUpstream, len(in.SidecarService.Proxy.Upstreams))
+			for i, p := range in.SidecarService.Proxy.Upstreams {
+				upstreams[i] = &structs.ConsulUpstream{
```

Comment:
It just occurred to me we don't need to store a pointer to upstreams and should just have a slice of values instead: `[]structs.ConsulUpstream`

It's what Consul seems to do: https://godoc.org/github.com/hashicorp/consul/api#ConnectProxyConfig

Not a blocker for this PR, but I wouldn't mind you sneaking it in :grin: